core/src/main/java/org/mule/runtime/core/internal/event/DefaultEventContext.java

```diff
@@ -345,6 +350,11 @@ public String getId() {
       return id;
     }
 
+    @Override
+    public String getRootId() {
+      return this.parent.getRootId();
```

Comment:
you have access to the root ctx from here (this.root) to avoid this second recursion